File: content/snippets/scala2/design_patterns_visitor.scala
// Visitor
trait Visitor {
    def visitCircle(c: Circle): Unit
    def visitRectangle(r: Rectangle): Unit
}

trait Element {
    def accept(visitor: Visitor): Unit
}

class Circle extends Element {
    def accept(visitor: Visitor): Unit = {
        visitor.visitCircle(this)
    }
    
    def draw(): Unit = println("Drawing circle")
}

class Rectangle extends Element {
    def accept(visitor: Visitor): Unit = {
        visitor.visitRectangle(this)
    }
    
    def draw(): Unit = println("Drawing rectangle")
}

class DrawVisitor extends Visitor {
    def visitCircle(c: Circle): Unit = {
        c.draw()
    }
    
    def visitRectangle(r: Rectangle): Unit = {
        r.draw()
    }
}

object Main extends App {
    val circle = new Circle()
    val rectangle = new Rectangle()
    val visitor = new DrawVisitor()
    
    circle.accept(visitor)
    rectangle.accept(visitor)
}

File: content/snippets/typescript/design_patterns_visitor.ts
interface IVisitor {
    visitCircle(c: Circle): void;
    visitRectangle(r: Rectangle): void;
}

interface IElement {
    accept(visitor: IVisitor): void;
}

class Circle implements IElement {
    accept(visitor: IVisitor): void {
        visitor.visitCircle(this);
    }
    
    draw(): void {
        console.log("Drawing circle");
    }
}

class Rectangle implements IElement {
    accept(visitor: IVisitor): void {
        visitor.visitRectangle(this);
    }
    
    draw(): void {
        console.log("Drawing rectangle");
    }
}

class DrawVisitor implements IVisitor {
    visitCircle(c: Circle): void {
        c.draw();
    }
    
    visitRectangle(r: Rectangle): void {
        r.draw();
    }
}

const circle = new Circle();
const rectangle = new Rectangle();
const visitor = new DrawVisitor();

circle.accept(visitor);
rectangle.accept(visitor);
